### PR TITLE
Mar 45/feat/search integration

### PR DESCRIPTION
**Pull request by shadrxcc**
# Mar 45/feat/search integration

> Fixes Issue

# Changes proposed

I have made the following changes to improve the search functionality:

- Stopped saving search results in local storage.
- Implemented the retrieval of search results directly from the server using search parameters.

# Check List
- [x] My code follows the code style of this project.
- [x] This PR does not contain plagiarized content.
- [x] The title and description of the PR is clear and explains the approach.
- [x] I am making a pull request against the **main branch** (left side).
- [x] My commit messages styles match our requested structure.
- [x] My code additions will fail neither code linting checks nor unit tests.
- [x] I am only making changes to files I was requested to.

### CODE REVIEW

File: http/api/searchPost.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import { ProductResult } from "../../@types";
+
+export const searchPosts = async (searchValue: string) => {
+    try {
+      const response = await fetch(
```
Comment (ickynavigator) [inline http/api/searchPost.ts]: shouldn't you use the axios package instead?

File: http/api/searchPost.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
+import { ProductResult } from "../../@types";
+
+export const searchPosts = async (searchValue: string) => {
```
Comment (ickynavigator) [inline http/api/searchPost.ts]: shouldn't this function(and file be called search products)

File: pages/marketplace/search/[query].tsx [hunk 3]
```diff
@@ -1,38 +1,46 @@
...
 import Error from '@modules/marketplace/component/landingpageerror/ErrorPage';
 import CategoryLayout from '@modules/marketplace/component/layout/category-layout';
+import { useRouter } from 'next/router';
+import { searchPosts } from '../../../http/api/searchPost';
 
 export default function Index() {
   const [results, setResults] = useState<ProductResult[]>([]);
-  const searchquery = typeof window !== 'undefined' ? localStorage.getItem('keyword') : null;
+  const router = useRouter();
+  const { query } = router.query;
```
Comment (ickynavigator) [inline pages/marketplace/search/[query].tsx]: destructure directly from the hook instead of this

File: pages/marketplace/search/[query].tsx [hunk 4]
```diff
@@ -1,38 +1,46 @@
 'use client';
```
Comment (ickynavigator) [inline pages/marketplace/search/[query].tsx]: don't switch this page to use a slug

use the query parameter instead

File: pages/marketplace/search/[query].tsx [hunk 5]
```diff
@@ -1,38 +1,46 @@
...
-      if (getresult) {
-        const result = JSON.parse(getresult);
-        setResults(result);
-      }
+    if (searchQuery) {
+      const fetchData = async () => {
+        try {
+          const results = await searchPosts(searchQuery);
+          setResults(results);
+          console.log(results)
```
Comment (ickynavigator) [inline pages/marketplace/search/[query].tsx]: remove the log

File: pages/marketplace/search/[query].tsx [hunk 6]
```diff
@@ -1,38 +1,46 @@
...
+        } catch (error) {
+          console.error(error);
```
Comment (ickynavigator) [inline pages/marketplace/search/[query].tsx]: this is not a valid error notification

File: pages/marketplace/search/[query].tsx [hunk 7]
```diff
@@ -1,38 +1,46 @@
...
+      fetchData();
     }
-  }, [results]);
+  }, [searchQuery]);
 
   return (
     <>
-      {results?.length === 0 ? (
+    {results?.length === 0 ? (
         <Error />
```
Comment (ickynavigator) [inline pages/marketplace/search/[query].tsx]: length 0 does not mean you should throw an error

show that the query returned nothing instead and another case to check if there was an error

File: components/Navbars/TopBar.tsx [hunk 8]
```diff
@@ -82,33 +83,33 @@ function TopBar(props: { activePage: string; showDashBorad: boolean }) {
...
   const handleSearch = async (e: React.KeyboardEvent) => {
     e.preventDefault();
     if (e.key === 'Enter' && dropDown === 'Marketplace') {
       try {
         const results = await searchPosts(searchQuery);
         setSearchResults(results);
-        localStorage.setItem('keyword', searchQuery);
-        localStorage.setItem('search_result', JSON.stringify(results));
-        router.push(`/marketplace/search?searchQuery=${searchQuery}`);
+        // localStorage.setItem('keyword', searchQuery);
```
Comment (ickynavigator) [inline components/Navbars/TopBar.tsx]: remove the comments

File: components/Navbars/TopBar.tsx [hunk 9]
```diff
@@ -82,33 +83,33 @@ function TopBar(props: { activePage: string; showDashBorad: boolean }) {
...
+  //     throw new Error(`Failed to fetch posts: ${error}`);
+  //   }
+  // };
 
   const handleSearch = async (e: React.KeyboardEvent) => {
     e.preventDefault();
     if (e.key === 'Enter' && dropDown === 'Marketplace') {
       try {
         const results = await searchPosts(searchQuery);
         setSearchResults(results);
```
Comment (ickynavigator) [inline components/Navbars/TopBar.tsx]: the query shouldn't be made in this file at all

**Review by ickynavigator — CHANGES_REQUESTED** [review]

**Comment by shadrxcc** [comment]
implemented changes as requested

File: http/api/searchProducts.ts [hunk 10]
```diff
@@ -0,0 +1,27 @@
+import axios from 'axios';
+import { ProductResult } from '../../@types';
+
+const MARKETPLACE_URL = `https://coral-app-8bk8j.ondigitalocean.app/api/product-retrieval/`;
+const axiosSearchInstance = axios.create({
```
Comment (ickynavigator) [inline http/api/searchProducts.ts:5]: you guys don't have an instance already defined?

File: http/api/searchProducts.ts [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+export const searchProducts = async (searchValue: string) => {
+  try {
+    const response = await axiosSearchInstance.get(`?search=${searchValue}`);
+
+    if (response.status !== 200) {
+      throw new Error('Network response was not ok');
+    }
+
+    const products: ProductResult[] = response.data;
+    const searchResults = products.filter((product) => product.name.toLowerCase().includes(searchValue.toLowerCase()));
```
Comment (ickynavigator) [inline http/api/searchProducts.ts]: why're you filtering from your end when the backend is doing that

Comment (shadrxcc) [reply http/api/searchProducts.ts]: they are but there's a different endpoint for filtering. the endpoint i'm using currently is '/product-retrieval'

File: http/api/searchProducts.ts (hunk 10 above)
Comment (shadrxcc) [reply http/api/searchProducts.ts:5]: there is in the axios.ts file. used it and got cors policy error so i created one in the meantime.

**Review by shalomtaiwo — APPROVED** [review]

File: http/api/searchProducts.ts (hunk 10 above)
Comment (ickynavigator) [reply http/api/searchProducts.ts:5]: alright

**Review by ickynavigator — APPROVED** [review]

**Pull request merged**